### PR TITLE
feat: introduce `set_parametes` and `set_initial_conds` functions

### PR DESCRIPTION
**Pull request by jia200x**
## Contribution description

This PR introduces 2 new functions for setting model parameters and initial conditions (see https://github.com/open-sir/open-sir/pull/61#issuecomment-614268738)

The old `set_params` function is still available, but added a deprecation note. We should migrate all examples to these new functions.

I also added some tests.

## Testing procedure

`pipenv run doc` and `pipenv run test`

## Related issues/PRs
Depends on #61

### CODE REVIEW

File: opensir/models/model.py [hunk 1]
```diff
@@ -28,8 +27,8 @@ class Model:
     """ Base model definition """
 
     CSV_ROW = []
-    NUM_PARAMS = 4
-    NUM_IC = 4
+    PARAMS = []
```
Comment (sasalatart) [inline opensir/models/model.py:30]: I like this. We avoid magic numbers :)

File: opensir/models/model.py [hunk 2]
```diff
@@ -57,18 +56,31 @@ def _model(self):
     def set_params(self, p, initial_conds):
         """ Set model parameters.
         Args:
-            p (list): parameters of the model. The parameters units are 1/day,
+            p (dict or list): parameters of the model. The parameters units are 1/day,
                       and should be >= 0.
```
Comment (sasalatart) [inline opensir/models/model.py:60]: Is this indentation ok? It does not seem to be consistent with `initial_conds`

File: opensir/models/model.py [hunk 3]
```diff
@@ -194,3 +206,10 @@ def function_handle(t, *par_fit, pop=population):
     def _update_ic(self):
         """ updates initial conditions if necessary """
         return self
+
+def _validate_params(arr, num_params):
+        if np.isnan(arr).any() or len(arr) != num_params:
+            raise Model.InvalidNumberOfParametersError
+
+        if sum(arr < 0) > 0:
```
Comment (sasalatart) [inline opensir/models/model.py:214]: Sorry, I am not familiar enough with Python. What is this line doing exactly?

I know you are adding through the array, but I am curious about the `arr < 0` part.

File: opensir/models/sir.py [hunk 4]
```diff
@@ -59,6 +63,13 @@ def set_params(self, p, initial_conds):
                 which is consistent with the mathematical description
                 of the SIR model.
 
+                If a list is used, the order of initial conditions is
+                [n_S0, n_I0, n_R0]
+
+        Deprecated:
+            This function is deprecated and will be removed soon.
```
Comment (sasalatart) [inline opensir/models/sir.py:70]:
```suggestion
            This function is deprecated and will be removed in future releases.
```

File: opensir/models/model.py (hunk 1 above)
Comment (felipehuerta17) [reply opensir/models/model.py:30]: I like avoiding magic numbers in the model mother class, but let's keep it in the sir and sirx models as the magic numbers are natural to the systems.

File: opensir/models/model.py (hunk 2 above)
Comment (felipehuerta17) [reply opensir/models/model.py:60]: Then initial_conds should change, as a tabbed second line is required to display well sphinx docs

File: opensir/models/model.py (hunk 3 above)
Comment (felipehuerta17) [reply opensir/models/model.py:214]: So arr < 0 yields a boolean array with 1 for each element of arr that's non-zero. Summing over that will produce the number of negative parameters. Then, the exception is raised if any of the elements of arr is negative.

**Comment by jia200x** [comment]
done! also rebased to lates master. I didn't remove the `Blocked` label because there are some fixups. As soon as this gets the final approve, I will squash and force push.

**Comment by felipehuerta17** [comment]
@jia200x fixups looks good! make sure that you lint and black_check as it is failing the CircleCI atm.

**Comment by felipehuerta17** [comment]
Actually this error is the best thing that it could have happened. The Ealing data had a terrible error:
```python
Ealing_data = [8, 18, 20, 18, 21, 42, 53, 54, 80, 97, 106, 123, 136, 165, 209, 241] # N_of infected
```

Non monotonically increasing number of infected!

**Pull request merged**